akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/ActorPublisherDocTest.java

```diff
@@ -0,0 +1,153 @@
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ActorPublisherDocTest {
+
+  static ActorSystem system;
+  
+
+  @BeforeClass
+  public static void setup() {
+    system = ActorSystem.create("ActorPublisherDocTest");
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    JavaTestKit.shutdownActorSystem(system);
+    system = null;
+  }
+  
+  final FlowMaterializer mat = ActorFlowMaterializer.create(system);
+  
+    //#job-manager
+    public static class JobManagerProtocol {
+      final public static class Job {
+        public final String payload;
+
+        public Job(String payload) {
+          this.payload = payload;
+        }
+
+      }
+
+      public static class JobAcceptedMessage {}
+      public static final JobAcceptedMessage JobAccepted = new JobAcceptedMessage () { 
+        @Override
+        public String toString() { return "JobAccepted"; }
+      };
+
+      static class JobDeniedMessage {}
+      public static final JobDeniedMessage JobDenied = new JobDeniedMessage () {
+        @Override
+        public String toString() { return "JobDenied"; }
+      };
+
+    }
+  
+    public static class JobManager extends AbstractActorPublisher<JobManagerProtocol.Job> {
+      
+      public static Props props() { return Props.create(JobManager.class); }
+
+      private final int MaxBufferSize = 100;
+      private final List<JobManagerProtocol.Job> buf = new ArrayList<>();
+
+      public JobManager() {
+        receive(ReceiveBuilder.
+          match(JobManagerProtocol.Job.class, job -> buf.size() == MaxBufferSize, job -> {
+            sender().tell(JobManagerProtocol.JobDenied, self());
+          }).
+          match(JobManagerProtocol.Job.class, job -> {
+            sender().tell(JobManagerProtocol.JobAccepted, self());
+
+            if (buf.isEmpty() && totalDemand() > 0)
+              onNext(job);
+            else {
+              buf.add(job);
+              deliverBuf();
+            }
+          }).
+          match(ActorPublisherMessage.Request.class, request -> deliverBuf()).
+          match(ActorPublisherMessage.Cancel.class, cancel -> context().stop(self())).
+          build());
+      }
+
+      void deliverBuf() {
+        if (totalDemand() > 0) {
+          /*
+           * totalDemand is a Long and could be larger than 
+           * what buf.splitAt can accept
+           */
+          if (totalDemand() <= Integer.MAX_VALUE) {
+            final List<JobManagerProtocol.Job> took =
+                buf.subList(0, Math.min(buf.size(), (int) totalDemand()));
+            took.forEach(this::onNext);
+            buf.removeAll(took);
+          } else {
+            final List<JobManagerProtocol.Job> took =
+                buf.subList(0, Math.min(buf.size(), Integer.MAX_VALUE));
+            took.forEach(this::onNext);
+            buf.removeAll(took);
+            deliverBuf(); // recurse
```

Comment:
shouldn't really be a problem here imo (would have to be a really huge buf), but a while loop is maybe safer to get people used to for those. Made it `while(true) + break`